README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->
'osrmr'
-------

'osrmr' is a wrapper around the OSRM API (<https://github.com/Project-OSRM/osrm-backend/blob/master/docs/http.md>). 'osrmr' works with API versions 4 and 5 and can handle servers that run locally as well as the osrm webserver.

Installation
------------

You can install 'osrmr' from `CRAN` or `github` with:

``` r
install.packages("osrmr")

# install.packages("devtools")
devtools::install_github("ims-fhs/osrmr")
```

Examples
--------

Access the OSRM with R using the onlinehost (webserver) of OSRM to:

-   generate coordinates from given coordinates, which are accessible by car with `nearest()`
-   calculate waytimes and more for specific routes from start- to end-destination with `viaroute()`
-   decode polylines using a polyline-specific precision with `decode_geom()`

``` r
library(osrmr)
nearest(lat = 47, lng = 8, api_version = 5, localhost = FALSE)
#>        lat      lng
#> 1 47.00008 8.003016
viaroute(lat1 = 47.1, lng1 = 8.1, lat2 = 46.9, lng2 = 8.3, instructions = FALSE,
         api_version = 5, localhost = FALSE)
#> [1] 2637.1

encoded_polyline_precision_5 <- rjson::fromJSON(file = "http://router.project-osrm.org/route/v1/driving/8.0997,47.1002;8.101110,47.10430?steps=false&geometries=polyline")$routes[[1]]$geometry
decode_geom(encoded_polyline_precision_5, precision = 5)
#>        lat     lng
#> 1 47.10020 8.09970
#> 2 47.10099 8.09952
#> 3 47.10161 8.10037
#> 4 47.10171 8.10066
#> 5 47.10215 8.10083
#> 6 47.10234 8.10098
#> 7 47.10287 8.10123
#> 8 47.10322 8.10125
#> 9 47.10430 8.10110
```

### Note

1.  In order to use the localhost of OSRM you need a local build on your device. For more Information on the localhost see <https://github.com/Project-OSRM/osrm-backend/wiki/Building-OSRM>.
2.  When using the localhost it's recommended to set the path of your local build as environment variable.

For more detailed Information about the components of the 'osrmr' package, check out the vignette.


NEWS.md
# osrmr 0.1.36

* Avoid calls to 'closeAllConnections()', as necessary change to remain on CRAN.

# osrmr 0.1.35

* Introduce an optional parameter 'timeout' for the functions 'viaroute' and 'nearest' that is preset to 0.001s. This prevents errors from server queue overflows.
* Improved error handling in 'viaroute' in order to comply with the CRAN policy on 'failing gracefully'.

# osrmr 0.1.29

* Set examples using OSRMR webserver to "dontrun" because of check errors due to flaky web server
* Bugfix in viaroute when using instructions
* Updated documentation

# osrmr 0.1.28

* Added a `NEWS.md` file to track changes to the package.
* Basic functionality as described in readme.md and vignette.





tests/testthat/helper-data.R
osrm_path <- "C:/OSRM_API5/"
map_name <- "switzerland-latest.osrm"
api_version <- 5
localhost <- FALSE


tests/testthat/test-basic_functions.R
context("viaroute")

test_that("viaroute finds routes", {
  # skip("Skip for development of tests")

  lat1 <- 47.168
  lng1 <- 8.117
  lat2 <- 46.978
  lng2 <- 8.335
  prec <- 0.02

  if(localhost) { run_server(osrm_path) }

  # Test the ABSOLUTE difference
  # expect_equal(viaroute(lat1, lng1, lat2, lng2, F, api_version, localhost),
  #              1560, tolerance = 300, scale = 1)
  # Test the RELATIVE difference is within prec
  # expect_equal(viaroute(lat1, lng1, lat2, lng2, F, 4, T),
  #              expected = 1812, tolerance = prec, scale = 1812)
  if(localhost) { quit_server() }
})

# # microbenchmark
# lat1 <- 52.503033
# lng1 <- 13.420526
# lat2 <- 52.516582
# lng2 <- 13.429290
# options(microbenchmark.unit = "s")
# suppressMessages(t9 <- summary(microbenchmark::microbenchmark(
#   sim911::osrm_viaroute(lat1, lng1, lat2, lng2),
#   times = 100L, unit = "ms"))$uq)
# expect_equal(t9 < t_400ms, T)
# # print(paste0("rjson_viaroute() took: ", round(t9, 1), " ms."))

# # osrm_viaroute: ........................................................... To osrm
# t6 <- summary(microbenchmark::microbenchmark(
#   osrm_viaroute(48, 8, 47, 7), times = 10L, unit = "ms"))$uq
# expect_equal(t6 < t_400ms, T)
# # print(paste0("osrm_viaroute needs: ", round(t6, 1), " ms."))


context("decoder")
test_that("decoder for consistent results", {
  # skip("Skip for development of tests")

  encoded_str <-"q|x}Gkuuq@u@Vw@ReAT_AJ}@Bs@CgAQgA[gAi@mAcAsDgDkByA}@k@_Ae@oAk@Ac@gAa@yAc@o@Qe@M}@Uo@IkAOsAGk@Bm@Fo@Js@RiAh@o@^s@h@_A|@_BhBuAhByA~BkHlMoCpDoAp@kA\\eCN}C?{D^_NdEoBr@uGtByIfD{@p@_A|@cAxAo@rAGHa@p@k@lAaBxDi@|@o@v@kAtAc@p@a@v@OPOFM?MMGMGM@EDO@UE]IQKIKCEAI@MHIJEPCT@PBPLTHDB@P@BAZ`@VXTNXHN?\\MTUp@_An@{@Zg@Tk@^cAlAsC^yABGfAwCz@yA|@mAz@u@v@s@rIeDtGuB|Bw@nMyDvDc@`DDfCOpAg@jAs@rCuDfHqMdBgCnAcBr@}@fAiAz@u@z@g@j@[v@Wv@Q^Eh@Id@Ab@?`@Bh@Dx@L`@H`B^l@RrC|@dA^~Ap@`Aj@t@f@hBtAhD~Ct@n@t@h@f@Xl@Tf@Pp@JhAFjA?bAKz@Q|@UjBo@j@Q`A[v@OdAM|@A|@BvBNbFh@tEf@~CXpCVxCPzD\\r@B`@?f@Ch@Gf@Ip@Sf@Sh@Yj@c@x@{@`AmAnBeCh@eAjAkC~EmI~@oCf@uB\\iBb@mDTuCPeJ_@qYOqCWgCa@gCq@cDk@iCi@sBa@iAi@aBY_AS_AQuAIuAC{A@uA@}ABuAH{ABc@Hw@RsARoAXuAXkA`@{ARq@Zy@Zs@\\s@h@_AlA_BxB{ChBgCzJaNrAiBlBkCtB}ClAgBfAaBpCuExAsCjA_CdAeC~@aC~@gCb@qAf@aBd@eBd@cB\\yA\\{@|@}DRkAVmAXoAXsB|@_HJg@Tk@V[VIHPLHL?HCJIDKBO@SCQISJSNe@H_@Dc@JwB@y@C[CYt@a@RMlBw@|Aq@XOBFBDFBD@DADCHd@TpA"
  # encoded <- strwrap(encoded_str, width=10000, simplify=T)

  geometry <- decode_geom(encoded_str, precision = 5)
  expect_equal(all(dim(geometry) == c(281, 2)), T)
  expect_equal(geometry$lat[1], 46.99097)
  expect_equal(geometry$lng[1], 8.30822)

  encoded_str_api_5 <- encoded_string_api_5
  geometry_api_5 <- decode_geom(encoded_str_api_5, precision = 6)
  expect_equal(all(dim(geometry_api_5) == c(35, 2)), T)
  expect_equal(geometry_api_5$lat[1], 47.100198)
  expect_equal(geometry_api_5$lng[1], 8.099703)
})



R/server.R
#' Start local OSRM server
#'
#' run_server() starts your local OSRM server by using a shell command (depending on
#' your OS). A local (pre-built) version of the OSRM-engine must be on your device.
#' (https://github.com/Project-OSRM/osrm-backend/wiki/Building-OSRM).
#'
#' To start the server, it is necessary to know its location. If it was installed
#' in C:/OSRM_APIx, it is easiest to set an environment variable which points to
#' the folder via Sys.setenv(). Note: You need to set the variable in each session.
#'
#' @param map_name A character (name of your pre-built map - e.g. "switzerland-latest.osrm")
#' @param osrm_path A string pointing to your local OSRM installation. Default is the environment variable "OSRM_PATH".
#'
#' @return error_code A character
#' @export
#' @examples
#' \dontrun{
#' Sys.setenv("OSRM_PATH"="C:/OSRM_API4")
#' osrmr::run_server("switzerland-latest.osrm")
#' # 0
#' Sys.setenv("OSRM_PATH"="C:/OSRM_API5")
#' osrmr::run_server("switzerland-latest.osrm")
#' # 0
#' Sys.unsetenv("OSRM_PATH")}
run_server <- function(map_name, osrm_path = Sys.getenv("OSRM_PATH")){
  wd <- getwd()
  setwd(osrm_path)

  if (.Platform$OS.type == "windows") {
    error_code <- shell(paste0("osrm-routed ", map_name, " >nul 2>nul"), wait = F)
  } else {
    error_code <- system(paste0(
      osrm_path, "osrm-routed ", map_name), wait = F)
  }

  Sys.sleep(3) # OSRM needs time
  setwd(wd)
  return(error_code)
}

#' Quit local OSRM server
#'
#' quit_server() quits your local OSRM server by using a taskkill shell command (depending on
#' your OS).
#'
#' @return NULL
#' @export
#' @examples
#' \dontrun{
#' osrmr::quit_server()
#' # NULL}
quit_server <- function() {
  if (.Platform$OS.type == "windows") {
    shell("TaskKill /F /IM osrm-routed.exe >nul 2>nul")
  } else {
    system("pkill osrm-routed")
  }
  return(NULL)
}


#' server_address() returns the URL address of the OSRM localhost or OSRM webserver,
#' depending on the value of the variable 'use_localhost'.
#' This is an internal function. The address is used as a part of a OSRM server-request.
#'
#' @param use_localhost A logical, indicating whether to use the localhost or not.
#'
#' @return character, the address of an OSRM server
#'
#' @examples
#' osrmr:::server_address(TRUE)
#' # [1] "http://localhost:5000"
#' osrmr:::server_address(FALSE)
#' # [1] "http://router.project-osrm.org"
server_address <- function(use_localhost) {
  if (use_localhost == T) {
    address <- "http://localhost:5000"
  } else {
    address <- "http://router.project-osrm.org"
  }
  return(address)
}


#' Run one server request for OSRM (online- or localhost)
#'
#' In order to fail gracefully, this function handles errors and warnings if the asked
#' server (online- or localhost) doesn't work properly. In this case the error message is
#' returned and connections are closed using base::closeAllConnections().
#'
#' If the asked server doesn't react within 1 second, a warning is thrown using
#' R.utils::withTimeout(..., timeout = 1)
#'
#' @param request A character
#'
#' @return A list. The dimension of the list depends on the request and whether the server reacted
#' properly or not.
make_request <- function(request) {
  R.utils::withTimeout({
    res <- tryCatch(
      {
        rjson::fromJSON(file = request)
      },
      error = function(cond) {
        message("Error:  OSRM Server doesn't react.")
        stop(cond)
      },
      warning = function(cond) {
        message("Warning: OSRM Server doesn't react.")
        stop(cond)
      }
    )},
    timeout = 1, onTimeout = "warning")
  return(res)
}




R/decoder.R
#' Transform encoded polylines to lat-lng data.frame.
#'
#' decode_geom() uses a decoding algorithm to decode polylines. (http://stackoverflow.com/questions/32476218/how-
#' to-decode-encoded-polylines-from-osrm-and-plotting-route-geometry)
#'
#' @param encoded A character containing encoded polylines
#' @param precision A numeric (either 5 or 6) to specify the precision of [lat,lng] encoding.
#' (OSRM API v4 used precision 5 with "polyline", OSRM API v5 uses precision 6 with "polyline6")
#'
#' @return data.frame with lat and lng
#' @export
#'
#' @examples
#' decoded_api_4 <- decode_geom(osrmr::encoded_string_api_4, precision = 5)
#' decoded_api_5 <- decode_geom(osrmr::encoded_string_api_5, precision = 6)
#' decoded_api_4[1:3,]
#' #        lat     lng
#' # 1 47.10020 8.09970
#' # 2 47.09850 8.09207
#' # 3 47.09617 8.09118
#' decoded_api_5[1:3,]
#' #        lat      lng
#' # 1 47.10020 8.099703
#' # 2 47.09850 8.092074
#' # 3 47.09617 8.091181
#' assertthat::assert_that(all.equal(decoded_api_4, decoded_api_5, tolerance = 1e-6))
decode_geom <- function(encoded, precision = stop("a numeric, either 5 or 6")) {
  if (precision == 5) {
    scale <- 1e-5
  } else if (precision == 6) {
    scale <- 1e-6
  } else {
    stop("precision not set to 5 or 6")
  }
  len = stringr::str_length(encoded)
  encoded <- strsplit(encoded, NULL)[[1]]
  index = 1
  N <- 100000
  df.index <- 1
  array = matrix(nrow = N, ncol = 2)
  lat <- dlat <- lng <- dlnt <- b <- shift <- result <- 0

  while (index <= len) {
    shift <- result <- 0
    repeat {
      b = as.integer(charToRaw(encoded[index])) - 63
      index <- index + 1
      result = bitops::bitOr(result, bitops::bitShiftL(bitops::bitAnd(b, 0x1f), shift))
      shift = shift + 5
      if (b < 0x20) break
    }
    dlat = ifelse(bitops::bitAnd(result, 1),
                  -(result - (bitops::bitShiftR(result, 1))),
                  bitops::bitShiftR(result, 1))
    lat = lat + dlat;

    shift <- result <- b <- 0
    repeat {
      b = as.integer(charToRaw(encoded[index])) - 63
      index <- index + 1
      result = bitops::bitOr(result, bitops::bitShiftL(bitops::bitAnd(b, 0x1f), shift))
      shift = shift + 5
      if (b < 0x20) break
    }
    dlng = ifelse(bitops::bitAnd(result, 1),
                  -(result - (bitops::bitShiftR(result, 1))),
                  bitops::bitShiftR(result, 1))
    lng = lng + dlng

    array[df.index,] <- c(lat = lat * scale, lng = lng * scale)
    df.index <- df.index + 1
  }

  geometry <- data.frame(array[1:df.index - 1,])
  names(geometry) <- c("lat", "lng")
  return(geometry)
}



R/nearest.R
#' nearest accessible position
#'
#' nearest() calculates the nearest position to the given coordinates which can be accessed by car.
#' The coordinate-standard is WGS84. Attention: The OSRM API v4 is only working locally, but
#' not with the 'OSRM' webserver.
#'
#' @param lat, A numeric (-90 < lat < 90)
#' @param lng, A numeric (-180 < lng < 180)
#' @param api_version, A numeric (either 4 or 5)
#' @param localhost, A logical (TRUE = localhost is used, FALSE = onlinehost is used)
#' @param timeout A numeric indicating the timeout between server requests (in order to prevent queue overflows). Default is 0.001s.
#'
#' @return A data.frame with lat and lng
#' @export
#'
#' @examples
#' \dontrun{
#' osrmr::nearest(47,9, 5, FALSE)
#'
#' Sys.setenv("OSRM_PATH_API_5"="C:/OSRM_API5")
#' osrmr::run_server(Sys.getenv("OSRM_PATH_API_5"), "switzerland-latest.osrm")
#' osrmr::nearest(47,9, 5, TRUE)
#' osrmr::quit_server()
#' Sys.unsetenv("OSRM_PATH_API_5")
#'
#' Sys.setenv("OSRM_PATH_API_4"="C:/OSRM_API4")
#' osrmr::run_server(Sys.getenv("OSRM_PATH_API_4"), "switzerland-latest.osrm")
#' osrmr::nearest(47,9, 4, TRUE)
#' osrmr::quit_server()
#' Sys.unsetenv("OSRM_PATH_API_4")}
nearest <- function(lat, lng, api_version = 5, localhost = F, timeout = 0.001) {
  assertthat::assert_that(api_version %in% c(4,5))

  address <- server_address(localhost)
  Sys.sleep(timeout)
  if (api_version == 4) {
    nearest <- nearest_api_v4(lat, lng, address)
  } else {
    nearest <- nearest_api_v5(lat, lng, address)
  }
  return(nearest)
}


#' nearest accessible position for OSRM API v4
#'
#' nearest_api_v4() calculates the nearest position to the given coordinates which can be accessed
#' by car with the OSRM API 4. The coordinate-standard is WGS84. Attention: The OSRM API v4
#' is only working locally, but not with the 'OSRM' webserver.
#'
#' @param lat, A numeric (-90 < lat < 90)
#' @param lng, A numeric (-180 < lng < 180)
#' @param address, A character specifying the serveraddress (local or online)
#'
#' @return A data.frame with lat and lng
#'
#' @examples
#' \dontrun{
#' Sys.setenv("OSRM_PATH_API_4"="C:/OSRM_API4")
#' osrmr::run_server(Sys.getenv("OSRM_PATH_API_4"), "switzerland-latest.osrm")
#' osrmr:::nearest_api_v4(47,9, osrmr:::server_address(TRUE))
#' osrmr::quit_server()
#' Sys.unsetenv("OSRM_PATH_API_4")}
nearest_api_v4 <- function(lat, lng, address) {
  request <- paste(address, "/nearest?loc=", lat, ",", lng, sep = "", NULL)
  nearest <- make_request(request)$mapped_coordinate
  nearest <- data.frame(
    lat = nearest[1],
    lng = nearest[2]
  )
  return(nearest)
}

#' nearest accessible position for OSRM API v5
#'
#' nearest_api_v5() calculates the nearest position to the given coordinates which can be accessed
#' by car with the OSRM API v5. The coordinate-standard is WGS84.
#'
#' @param lat, A numeric (-90 < lat < 90)
#' @param lng, A numeric (-180 < lng < 180)
#' @param address, A character specifying the serveraddress (local or online)
#'
#' @return A data.frame with lat and lng

#' @examples
#' \dontrun{
#' osrmr:::nearest_api_v5(47,9, osrmr:::server_address(FALSE))
#' Sys.setenv("OSRM_PATH_API_5"="C:/OSRM_API5")
#' osrmr::run_server(Sys.getenv("OSRM_PATH_API_5"), "switzerland-latest.osrm")
#' osrmr:::nearest_api_v5(47,9, osrmr:::server_address(TRUE))
#' osrmr::quit_server()
#' Sys.unsetenv("OSRM_PATH_API_5")}
nearest_api_v5 <- function(lat, lng, address) {
  request <-paste(address, "/nearest/v1/driving/", lng, ",", lat, "?number=1", sep = "", NULL)
  nearest <- make_request(request)$waypoints[[1]]$location
  nearest <- data.frame(
    lat = nearest[2],
    lng = nearest[1]
  )
  return(nearest)
}




R/viaroute.R
#' travel time or full information of a route
#'
#' For a given start- and end-destination, viaroute() calculates route informations using OSRM.
#' OSRM chooses the nearest point which can be accessed by car for the start- and end-destination.
#' The coordinate-standard is WGS84.
#' Attention: The OSRM API-4 is only working locally, but not with the onlinehost.
#'
#' @param lat1 A numeric (-90 < lat1 < 90) -> start-destination
#' @param lng1 A numeric (-180 < lng1 < 180) -> start-destination
#' @param lat2 A numeric (-90 < lat2 < 90) -> end-destination
#' @param lng2 A numeric (-180 < lng2 < 180) -> end-destination
#' @param instructions A logical. If FALSE, only the traveltime (in seconds, as numeric) will be returned.
#'  If TRUE, more details of the route are returned (as list).
#' @param api_version A numeric (either 4 or 5)
#' @param localhost A logical (TRUE = localhost is used, FALSE = onlinehost is used)
#' @param timeout A numeric indicating the timeout between server requests (in order to prevent queue overflows). Default is 0.001s.
#'
#' @return a numeric or a list (depending on instructions)
#' @export
#'
#' @examples
#' # direct examples of the online API
#' \dontrun{
#' #' link <- "http://router.project-osrm.org/route/v1/driving/8.1,47.1;8.3,46.9?steps=false"
#' a <- rjson::fromJSON(file = link)
#'
#' # example with onlinehost API5
#' osrmr:::viaroute(47.1, 8.1, 46.9, 8.3, FALSE, 5, FALSE)
#'
#' # examples with localhost API4/API5
#' Sys.setenv("OSRM_PATH"="C:/OSRM_API4")
#' osrmr::run_server("switzerland-latest.osrm")
#' osrmr::viaroute(47.1, 8.1, 46.9, 8.3, FALSE, 4, TRUE)
#' osrmr::quit_server()
#' Sys.unsetenv("OSRM_PATH")
#'
#' Sys.setenv("OSRM_PATH"="C:/OSRM_API5")
#' osrmr::run_server("switzerland-latest.osrm")
#' osrmr::viaroute(47.1, 8.1, 46.9, 8.3, FALSE, 5, TRUE)
#' osrmr::quit_server()
#' Sys.unsetenv("OSRM_PATH")}
viaroute <- function(lat1, lng1, lat2, lng2, instructions, api_version, localhost, timeout = 0.001) {
  assertthat::assert_that(api_version %in% c(4,5))
  address <- server_address(localhost)

  Sys.sleep(timeout)
  if (api_version == 4) {
    viaroute_api_v4(lat1, lng1, lat2, lng2, instructions, address)
  } else {
    viaroute_api_v5(lat1, lng1, lat2, lng2, instructions, address)
  }

}


#' travel time or full information of a route for OSRM API 4
#'
#' For a given start- and end-destination, viaroute() calculates route informations using OSRM API 4.
#' OSRM chooses the nearest point which can be accessed by car for the start and destination.
#' The coordinate-standard is WGS84.
#' Attention: The OSRM API-4 is only working locally, but not with the onlinehost.
#'
#' @param lat1 A numeric (-90 < lat1 < 90) -> start-destination
#' @param lng1 A numeric (-180 < lng1 < 180) -> start-destination
#' @param lat2 A numeric (-90 < lat2 < 90) -> end-destination
#' @param lng2 A numeric (-180 < lng2 < 180) -> end-destination
#' @param instructions A logical. If FALSE, only the traveltime (in seconds, as numeric) will be returned.
#'  If TRUE, more details of the route are returned (as list).
#' @param address A character specifying the serveraddress (local or online)
#'
#' @return a numeric or a list (depending on parameter instructions)
#'
#' @examples
#' \dontrun{
#' Sys.setenv("OSRM_PATH"="C:/OSRM_API4")
#' osrmr::run_server("switzerland-latest.osrm")
#' osrmr:::viaroute_api_v4(47,9,48,10, FALSE, osrmr:::server_address(TRUE))
#' osrmr::quit_server()
#' Sys.unsetenv("OSRM_PATH")}
viaroute_api_v4 <- function(lat1, lng1, lat2, lng2, instructions, address) {
  request <- paste(address, "/viaroute?loc=",
                   lat1, ',', lng1, '&loc=', lat2, ',', lng2, sep = "", collapse = NULL)
  res <- make_request(request)

  if (!instructions) {
    if (!res$status == 207) {
      return(res$route_summary$total_time)
    } else {
      t_guess <- 16*60
      warning("Route not found: ", paste(lat1, lng1, lat2, lng2, collapse = ", "),
              ". Travel time set to ", t_guess/60 , " min.")
      return(t_guess) # Guess a short walk of 2 minutes.
    }
  } else {
    return(res)
  }
}



#' travel time or full information of a route for OSRM API 5
#'
#' For a given start- and end-destination, viaroute() calculates route informations using OSRM API 5.
#' OSRM chooses the nearest point which can be accessed by car for the start and destination.
#' The coordinate-standard is WGS84.
#' Attention: The OSRM API-4 is only working locally, but not with the onlinehost.
#'
#' @param lat1 A numeric (-90 < lat1 < 90) -> start-destination
#' @param lng1 A numeric (-180 < lng1 < 180) -> start-destination
#' @param lat2 A numeric (-90 < lat2 < 90) -> end-destination
#' @param lng2 A numeric (-180 < lng2 < 180) -> end-destination
#' @param instructions A logical. If FALSE, only the traveltime (in seconds, as numeric) will be returned.
#'  If TRUE, more details of the route are returned (as list).
#' @param address A character specifying the serveraddress (local or online)
#'
#' @return a numeric or a list (depending on parameter instructions)
#'
#' @examples
#' \dontrun{
#' # example with onlinehost
#' osrmr:::viaroute_api_v5(47, 9, 48, 10 , FALSE, osrmr:::server_address(FALSE))
#'
#' # example with localhost
#' Sys.setenv("OSRM_PATH"="C:/OSRM_API5")
#' osrmr::run_server("switzerland-latest.osrm")
#' osrmr:::viaroute_api_v5(47, 9, 48, 10 , FALSE, osrmr:::server_address(TRUE))
#' osrmr::quit_server()
#' Sys.unsetenv("OSRM_PATH")}
viaroute_api_v5 <- function(lat1, lng1, lat2, lng2, instructions, address) {
  if (!instructions) {
    request <- paste(address, "/route/v1/driving/",
                     lng1, ",", lat1, ";", lng2, ",", lat2,
                     "?overview=false", sep = "", NULL)
  } else {
    request <- paste(address, "/route/v1/driving/",
                     lng1, ",", lat1, ";", lng2, ",", lat2,
                     "?overview=full", sep = "", NULL)
  }
  res <- make_request(request)

  assertthat::assert_that(assertthat::is.number(res$routes[[1]]$duration))

  if (!instructions) {
    if (res$code == "Ok") {
      return(res$routes[[1]]$duration)
    }
    else {
      t_guess <- 16*60
      warning("Route not found: ", paste(lat1, lng1, lat2, lng2, collapse = ", "),
              ". Travel time set to ", t_guess/60 , " min.")
    }
  } else {
    return(res)
  }
}

